include/revng/ABIAnalyses/Common.h

```diff
@@ -0,0 +1,213 @@
+
+  SmallVector<const GlobalVariable *, 1>
+  getRegistersWritten(const Instruction *) const;
+
+  SmallVector<const GlobalVariable *, 1>
+  getRegistersRead(const Instruction *) const;
+};
+
+template<typename LatticeElementMap, typename CoreLattice, typename KeyT>
+typename CoreLattice::LatticeElement
+getOrDefault(const LatticeElementMap &S, const KeyT &K) {
+  auto V = S.find(K);
+  if (V != S.end()) {
+    return S.lookup(K);
+  }
+  return CoreLattice::DefaultLatticeElement;
+}
+
+template<typename LatticeElementMap, typename CoreLattice>
+LatticeElementMap
+combineValues(const LatticeElementMap &Lh, const LatticeElementMap &Rh) {
+
+  LatticeElementMap New = Lh;
+  for (const auto &[Reg, S] : Rh) {
+    New[Reg] = CoreLattice::
+      combineValues(getOrDefault<LatticeElementMap, CoreLattice>(New, Reg),
+                    getOrDefault<LatticeElementMap, CoreLattice>(Rh, Reg));
+  }
+  return New;
+}
+
+template<typename LatticeElementMap, typename CoreLattice>
+bool isLessOrEqual(const LatticeElementMap &Lh, const LatticeElementMap &Rh) {
+  for (auto &[Reg, S] : Lh) {
+    if (!CoreLattice::isLessOrEqual(getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Lh, Reg),
+                                    getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Rh, Reg))) {
+      return false;
+    }
+  }
+  for (auto &[Reg, S] : Rh) {
+    if (!CoreLattice::isLessOrEqual(getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Lh, Reg),
+                                    getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Rh, Reg))) {
+      return false;
+    }
+  }
+  return true;
+}
+
+inline bool ABIAnalysis::isABIRegister(const Value *V) const {
+  if (const auto &G = dyn_cast<GlobalVariable>(V)) {
+    if (ABIRegisters.count(G) != 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+inline bool isCallSiteBlock(const BasicBlock *B) {
+  if (auto *C = dyn_cast<CallInst>(&*B->getFirstInsertionPt())) {
+    if (C->getCalledFunction()->getName().contains("precall_hook")) {
+      return true;
+    }
+  }
+  return false;
+}
+
+inline const Instruction *getPreCallHook(const BasicBlock *B) {
+  if (isCallSiteBlock(B)) {
+    return &*B->getFirstInsertionPt();
+  }
+  return nullptr;
+}
+
+inline const Instruction *getPostCallHook(const BasicBlock *B) {
+  if (isCallSiteBlock(B)) {
+    return B->getTerminator()->getPrevNode();
+  }
+  return nullptr;
+}
+
+inline TransferKind
+ABIAnalysis::classifyInstruction(const Instruction *I) const {
+  switch (I->getOpcode()) {
+  case Instruction::Store: {
+    auto S = cast<StoreInst>(I);
+    if (isABIRegister(S->getPointerOperand())) {
+      if (isCallSiteBlock(I->getParent())) {
+        return WeakWrite;
+      }
+      return Write;
+    }
+    break;
+  }
+  case Instruction::Load: {
+    auto L = cast<LoadInst>(I);
```

Comment:
Don't let `auto` deduce `*`
```suggestion
    auto *L = cast<LoadInst>(I);
```